xapian-applications/omega/handler_libextractor.cc

```diff
@@ -0,0 +1,133 @@
+/** @file handler_libextractor.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <extractor.h>
+#include <cstring>
+
+using namespace std;
+
+struct metadata {
+    string& title;
+    string& author;
+    string& keywords;
+    string& pages;
+
+    metadata(string& t, string& a, string& k, string& p)
+	    : title(t), author(a), keywords(k), pages(p) {}
+};
+
+/*
+* cls: passed as last parameter from EXTRACTOR_extract
+* plugin_name: name of the plugin
+* type: mime-type of file according to libextractor
+* format: format information about data
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
+static int
+process_metadata(void* cls,
+		 const char* plugin_name,
+		 enum EXTRACTOR_MetaType type,
+		 enum EXTRACTOR_MetaFormat format,
+		 const char* data_mime_type,
+		 const char* data,
+		 size_t data_len)
+{
+    struct metadata* md = (struct metadata*)cls;
+
+    switch (format) {
+	case EXTRACTOR_METAFORMAT_UTF8:
+	    break;
+
+	default:
+	    // specific encoding unknown
+	    // EXTRACTOR_METAFORMAT_UNKNOWN
+	    // EXTRACTOR_METAFORMAT_BINARY
+	    // EXTRACTOR_METAFORMAT_C_STRING
+	    return 0;
+    }
+
+    switch (type) {
+	case EXTRACTOR_METATYPE_BOOK_TITLE:
+	case EXTRACTOR_METATYPE_JOURNAL_NAME:
+	case EXTRACTOR_METATYPE_ORIGINAL_TITLE:
+	case EXTRACTOR_METATYPE_SUBJECT:
+	case EXTRACTOR_METATYPE_TITLE:
+	    md->title += string(data, data_len);
+	    md->title += " ";
```

Comment:
This creates a temporary `std::string` object to append the data, which will result in needlessly copying the data, and also in a heap allocation if the string is at all long (short strings are typically actually just stored within the `std::string` object itself).  This temporary object can be avoided by using the `std::string::append()` method - that has an overload form which takes a pointer and length:

```
md->title.append(data, data_len);
```

Also, this always results in a trailing space on the result - it'd be neater to only add it if there's already something in the variable, which is easy to do by first checking and adding the space if non-empty, then appending the data.

Reply:
I have made the necessary changes here.